### PR TITLE
Mostly fix 637

### PR DESCRIPTION
**Pull request by eric-wieser**
#637

This prevents the innards of `react` being entered more than once, which as a bonus seems to fix the mentioned issue

This leaves a lingering bug for tests which exit before yielding, but that's not nearly as useful as coroutines that do.

---

Continues from #787. Happy to squash these commits together if it makes the change clearer.

Again, diff is by far clearest with whitespace turned off:
image

### CODE REVIEW

File: tests/test_cases/test_cocotb/test_cocotb.py [hunk 1]
```diff
@@ -869,10 +870,54 @@ def test_exceptions():
...
+
     try:
         yield raise_soon()
     except ValueError:
         pass
     else:
         raise TestFailure("Exception was not raised")
+
+@cocotb.test()
+def test_stack_overflow(dut):
```
Comment (eric-wieser) [inline tests/test_cases/test_cocotb/test_cocotb.py]: This test is lifted from #625

File: cocotb/scheduler.py [hunk 2]
```diff
@@ -319,13 +320,36 @@ def begin_test(self, trigger=None):
...
+            self._is_reacting = True
+            try:
+                self._event_loop(trigger)
+            finally:
+                self._is_reacting = False
+
+
+    def _event_loop(self, trigger):
+        """
+        Run an event loop trigerred by the given trigger.
```
Comment (imphil) [inline cocotb/scheduler.py]: typo: triggered

File: cocotb/scheduler.py [hunk 3]
```diff
@@ -319,13 +320,36 @@ def begin_test(self, trigger=None):
...
+        """
+        Called when a trigger fires.
+
+        We ensure that we only start the event loop once, rather than
+        letting it recurse.
+        """
+        if self._is_reacting:
+            # queue up the trigger, the event loop will get to it
+            self._pending_triggers.append(trigger)
+        else:
```
Comment (imphil) [inline cocotb/scheduler.py]: An early return would be potentially nicer here, but I leave it up to you.

**Comment by imphil** [comment]
And am I correct that we can close #787 when we take this one?

**Comment by eric-wieser** [comment]
Yes 787 existed just in case you wanted to merge this in two pieces

File: cocotb/scheduler.py (hunk 3 above)
Comment (eric-wieser) [reply cocotb/scheduler.py]: Fine be me either way - changed

**Comment by imphil** [comment]
@eric-wieser did you mean to push an update to this one that somehow got lost?

**Comment by eric-wieser** [comment]
Good catch, Yes, pushed it over the other PR instead. I've closed that one now, and think I got the push correct this time.

**Comment by imphil** [comment]
Travis seems to be confused with this PR, retriggering didn't do anything to resolve it. Let's see how it does on master and back of if necessary.

Thanks @eric-wieser, these PRs are great!

**Pull request merged**